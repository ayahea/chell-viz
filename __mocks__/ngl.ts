/**
 * Mock implementations for the [NGL](https://github.com/arose/ngl) library.
 *
 * Our NGLComponent is, understandably, pretty coupled to the library so this is an attempt to mock the behaviors.
 *
 * Ideally it would be most beneficial if as much of the original ngl / automocked version could be used.
 * As of this writing, it is not 100% clear to me on how best to approach selectively mocking several
 * classes/methods of the 3rd party library while retaining most of the original functionality.
 *
 * https://facebook.github.io/jest/docs/en/manual-mocks.html
 */

import * as NGL from 'ngl';

const ngl = jest.genMockFromModule<typeof NGL>('ngl');

class MockStage {
  public events = new Map<string, (...args: any[]) => void>();
  public callbacks = new Array<(...args: any[]) => void>();
  public reprList: string[] = [];

  public mouseControls = {
    add: (eventName: string, callback: (...args: any[]) => void) => this.events.set(eventName, callback),
    run: (eventName: string, ...args: any[]) => {
      if (this.events.get(eventName)) {
        this.events.get(eventName)!(...args);
      }
    },
  };

  public signals = {
    clicked: {
      add: (callback: (...args: any[]) => void) => this.events.set('click', callback),
      dispatch: (...args: any[]) => {
        if (this.events.get('click')) {
          this.events.get('click')!(...args);
        }
      },
    },
  };

  constructor(canvas: HTMLElement) {
    return;
  }

  public addComponentFromObject = () => ({
    addRepresentation: (name: string, ...args: any[]) => {
      this.reprList.push(name);
      return name;
    },
    hasRepresentation: (name: string, ...args: any[]) => this.reprList.indexOf(name) !== -1,
    removeRepresentation: (name: string, ...args: any[]) => {
      this.reprList.splice(this.reprList.indexOf(name), 1);
    },

    stage: {
      mouseControls: this.mouseControls,
    },
  });
  public defaultFileRepresentation = (...args: any[]) => jest.fn();
  public dispose = () => jest.fn();
  public removeAllComponents = () => jest.fn();
}

(ngl.Stage as any).mockImplementation((canvas: HTMLCanvasElement) => {
  return new MockStage(canvas);
});

(ngl.Structure as any).mockImplementation(() => {
  return {
    getSequence: jest.fn(() => []),
    residueMap: {
      list: [],
    },
  };
});

(ngl.autoLoad as any) = jest.fn(
  (path: string) =>
    path.localeCompare('error/protein.pdb') === 0
      ? Promise.reject('Invalid NGL path.')
      : Promise.resolve('Mock NGL path.'),
);

// @ts-ignore
ngl.MouseActions = {
  CLICK: 'click',
  CLICK_PICK: 'clickPick',
  DOUBLE_CLICK: 'doubleClick',
  DRAG: 'drag',
  HOVER: 'hover',
  HOVER_PICK: 'hoverPick',
  SCROLL: 'scroll',
};

module.exports = ngl;
