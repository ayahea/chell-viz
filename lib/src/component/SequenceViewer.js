"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var SequenceViewer = /** @class */ (function (_super) {
    __extends(SequenceViewer, _super);
    function SequenceViewer(props) {
        return _super.call(this, props) || this;
    }
    SequenceViewer.prototype.render = function () {
        var _a = this.props, data = _a.data, selectedResNum = _a.selectedResNum;
        if (data) {
            var fullSeq = data.getSequence().join('');
            return (React.createElement("div", { id: "SequenceViewer" },
                this.renderFullSequence(fullSeq, selectedResNum),
                data.residueMap.list.map(function (value, index) {
                    if (selectedResNum &&
                        value.resname === data.residueMap.get(data.residueStore.residueTypeId[selectedResNum]).resname) {
                        return React.createElement("div", { key: index }, value.resname + " was selected!");
                    }
                    else {
                        return React.createElement("div", { key: index }, value.resname);
                    }
                })));
        }
        return null;
    };
    SequenceViewer.prototype.renderFullSequence = function (sequence, selectedResNum) {
        if (selectedResNum === void 0) { selectedResNum = -1; }
        return (React.createElement("div", { style: { width: 400, maxHeight: 400, wordWrap: 'break-word' } },
            sequence.substr(0, selectedResNum + 1),
            React.createElement("span", { style: { color: 'red', fontSize: 24 } }, sequence[selectedResNum]),
            sequence.substr(selectedResNum + 1)));
    };
    return SequenceViewer;
}(React.Component));
exports.SequenceViewer = SequenceViewer;
