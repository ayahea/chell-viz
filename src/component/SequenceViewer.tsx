import * as NGL from 'ngl';
import * as React from 'react';

export interface ISequenceViewerProps {
  data?: NGL.Structure;
  selectedResname?: string;
  onSeqHoverCallback?: (...args: any[]) => void;
  onSeqClickCallback?: (...args: any[]) => void;
}

export class SequenceViewer extends React.Component<ISequenceViewerProps, any> {
  private featureViewer: HTMLElement | null = null;

  constructor(props: any) {
    super(props);
  }

  public render() {
    const { data, selectedResname } = this.props;
    return (
      <div id="SequenceViewer">
        <div onMouseOver={this.props.onSeqHoverCallback}>{data && data.getSequence().slice(0, 50)}</div>
        {data &&
          data.residueMap.list.map((value, index) => {
            if (value.resname === selectedResname) {
              return <div key={index}>{`${value.resname} was selected!`}</div>;
            } else {
              return <div key={index}>{value.resname}</div>;
            }
          })}
        <div id="fv1" ref={el => (this.featureViewer = el)} style={{ height: 400, width: 400 }} />
      </div>
    );
  }
}
