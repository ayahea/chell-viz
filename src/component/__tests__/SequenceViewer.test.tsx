import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as NGL from 'ngl';
import * as React from 'react';

import SequenceViewer from '../SequenceViewer';

describe('SequenceViewer', () => {
  test('Should match existing snapshot when given no data.', () => {
    expect(toJson(shallow(<SequenceViewer data={undefined} />))).toMatchSnapshot();
  });

  test('Should match existing snapshot when given sample data.', () => {
    const mockData = new NGL.Structure();
    const component = shallow(<SequenceViewer data={mockData} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
