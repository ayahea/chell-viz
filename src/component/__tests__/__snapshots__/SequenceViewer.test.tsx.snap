// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceViewer Should match existing snapshot when given no data. 1`] = `""`;
